File: JavaScript_/BinaryTree/dfs.js
class Node{
  constructor(val){
    this.val =  val;
    this.left = null;
    this.right= null;
  }
  
}

const depthFirstValue = (root) =>{
  
  if (root === null)return [];
  
  const result = [];
  const stack = [root];
  while(stack.length > 0){
    const current = stack.pop()
    
   result.push(current.val);
   
    if (current.left)
      stack.push(current.left)
    if (current.right)
      stack.push(current.right)
      
  
    
  }
  return result;
  
}


const depthFirstValueRecurssive = (root) =>{
  
  if(root === null) return [];
  
  const leftValues =  depthFirstValueRecurssive(root.left);
  const rightValues = depthFirstValueRecurssive(root.right);
  
  return [root.val, ...leftValues, ...rightValues]
  
}


const a = new Node("A")
const b = new Node("B")
const c = new Node("C")
const d = new Node("D")
const e = new Node("E")
const f = new Node("F")

a.left = b;
a.right = c;
b.left = d ;
b.right = e ;
c.right = f;


console.log(
  depthFirstValue(a)
)

File: Python_/Binary_tree/dfs.py
class Node:
    def __init__(self, value):
        self.value = value
        self.left = None
        self.right = None


def DepthFirstSearch(root):
    # if root is empty
    if root == None:
        return []
    else:
        # for storing result
        result = []
        # for doing DFs
        stack = [root]
        # loop through until the stack is empty
        while len(stack) > 0:
            # pop current item
            current = stack.pop()
            # append it to result
            result.append(current.value)

            # check if right not exist if yes then append
            if current.right:
                stack.append(current.right)
            # check if left not exist if yes then append
            if current.left:
                stack.append(current.left)

    return result


def recurssiveDFS(root):
    if root == None:
        return []

    leftValues = recurssiveDFS(root.left)
    rightValues = recurssiveDFS(root.right)

    return [root.value, *leftValues, *rightValues]


a = Node("A")
b = Node("B")
c = Node("C")
d = Node("D")
e = Node("E")
f = Node("F")

a.left = b
a.right = c
b.left = d
b.right = e
c.right = f

#     a
#    /  \
#    b  c
#  /  \  \
#  d  e   f
# print(DepthFirstSearch(a))
print(recurssiveDFS(a))
